source/drivers/ST7735.cpp

```diff
@@ -0,0 +1,352 @@
+    }
+
+    work->srcPtr = (uint8_t *)src;
+    startTransfer((uint8_t *)dst - work->dataBuf);
+}
+
+void ST7735::sendColorsStep(ST7735 *st)
+{
+    ST7735WorkBuffer *work = st->work;
+
+    if (work->paletteTable) {
+        auto palette = work->paletteTable;
+        work->paletteTable = NULL;
+        memset(work->dataBuf, 0, sizeof(work->dataBuf));
+        uint8_t *base = work->dataBuf;
+        for (int i = 0; i < 16; ++i)
+        {
+            base[i] = (palette[i] >> 18) & 0x3f;
+            base[i + 32] = (palette[i] >> 10) & 0x3f;
+            base[i + 32 + 64] = (palette[i] >> 2) & 0x3f;
+        }
+        st->startRAMWR(0x2D);
+        st->spi.transfer(work->dataBuf, 128, NULL, 0);
+        st->cs.setDigitalValue(1);
+    }
+
+    if (work->x == 0)
+    {
+        st->startRAMWR();
+        work->x++;
+    }
+
+    // with the current image format in PXT the sendBytes cases never happen
+    unsigned align = (unsigned)work->srcPtr & 3;
+    if (work->srcLeft && align)
+    {
+        st->sendBytes(4 - align);
+    }
+    else if (work->srcLeft < 4)
+    {
+        if (work->srcLeft == 0)
+        {
+            st->cs.setDigitalValue(1);
+            Event(DEVICE_ID_DISPLAY, 100);
+        }
+        else
+        {
+            st->sendBytes(work->srcLeft);
+        }
+    }
+    else
+    {
+        st->sendWords((sizeof(work->dataBuf) / (3 * 4)) * 4);
+    }
+}
+
+void ST7735::startTransfer(unsigned size)
+{
+    spi.startTransfer(work->dataBuf, size, NULL, 0, (PVoidCallback)&ST7735::sendColorsStep, this);
+}
+
+void ST7735::startRAMWR(int cmd)
+{
+    if (cmd == 0)
+        cmd = ST7735_RAMWR;
+    cmdBuf[0] = cmd;
+    sendCmd(cmdBuf, 1);
+
+    dc.setDigitalValue(1);
+    cs.setDigitalValue(0);
+}
+
+void ST7735::sendDone(Event)
+{
+    // this executes outside of interrupt context, so we don't get a race
+    // with waitForSendDone
+    work->inProgress = false;
+    Event(DEVICE_ID_DISPLAY, 101);
+}
+
+void ST7735::waitForSendDone()
+{
+    if (work && work->inProgress)
+        fiber_wait_for_event(DEVICE_ID_DISPLAY, 101);
+}
+
+int ST7735::sendIndexedImage(const uint8_t *src, unsigned width, unsigned height, uint32_t *palette)
+{
+    if (!work)
+    {
+        work = new ST7735WorkBuffer;
+        memset(work, 0, sizeof(*work));
+        for (int i = 0; i < 256; ++i)
+            work->expPalette[i] = 0x1011 * (i & 0xf) | (0x110100 * (i>>4));
+        EventModel::defaultEventBus->listen(DEVICE_ID_DISPLAY, 100, this, &ST7735::sendDone);
+    }
+
+    if (work->inProgress)
+        return DEVICE_BUSY;
+
+    work->paletteTable = palette;
+
+    work->inProgress = true;
+    work->srcPtr = src;
+    work->width = width;
+    work->height = height;
+    work->srcLeft = ((width + 1) >> 1) * height;
+    work->x = 0;
+
+    sendColorsStep(this);
+
+    return DEVICE_OK;
+}
+
+// we don't modify *buf, but it cannot be in flash, so no const as a hint
+void ST7735::sendCmd(uint8_t *buf, int len)

```

Comment:
It seems like `sendCmd`, `sendBytes`, and `sendWords` should actually be the interfaces for the SPI abstraction?